src/bin/pgcopydb/ld_replay.c

```diff
@@ -249,3 +255,166 @@ stream_replay_line(void *ctx, const char *line, bool *stop)
+		return false;
+	}
+
+	fclose(txnBufferForRead);
+
+	return true;
+}
+
+
+/* streamReplayLineIntoBuffer is a callback function for the ReadFromStreamContext
+ * and read_from_stream infrastructure. It's called on each line read from a
+ * stream such as a unix pipe and buffers the line into a file until we encounter
+ * a COMMIT. This ensures that we can replay the whole transaction in one go
+ * and avoids deadlock when the BEGIN messages arrives before the existence of
+ * transation metadata file which is created by the transform process.
+ */
+static bool
+streamReplayLineIntoBuffer(void *ctx, const char *line, bool *stop)
+{
+	ReplayStreamCtx *replayCtx = (ReplayStreamCtx *) ctx;
+	StreamApplyContext *context = &(replayCtx->applyContext);
+
+	LogicalMessageMetadata metadata = { 0 };
+
+	if (!parseSQLAction((char *) line, &metadata))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	/* We don't use this metadata in this function */
+	free(metadata.jsonBuffer);
+
+	if (metadata.action == STREAM_ACTION_BEGIN)
+	{
+		if (replayCtx->isTranscationBuffering)
+		{
+			log_debug("Received %s when transaction is already "
+					  "in buffering mode, ignoring.", line);
+			fclose(replayCtx->txnBuffer);
+		}
+
+		sformat(replayCtx->currentTxnFileName, MAXPGPATH, "%s/%d.sql",
+				context->paths.dir, metadata.xid);
+
+		/*
+		 * Open current transaction file in w+ mode to truncate if the file
+		 * already exists.
+		 */
+		replayCtx->txnBuffer = fopen_with_umask(replayCtx->currentTxnFileName,
+												"w+",
+												O_RDWR | O_TRUNC | O_CREAT,
+												0644);
+		if (replayCtx->txnBuffer == NULL)
+		{
+			log_error("Failed to open transaction buffer file \"%s\": %m",
+					  replayCtx->currentTxnFileName);
+			return false;
+		}
+
+		/* Create as a symlink to the current transcation file */
+		if (!create_symbolic_link(replayCtx->currentTxnFileName,
+								  context->paths.txnlatestfile))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		fformat(replayCtx->txnBuffer, "%s\n", line);
+
+		replayCtx->isTranscationBuffering = true;
+	}
+	else if (metadata.action == STREAM_ACTION_COMMIT)
+	{
+		if (!replayCtx->isTranscationBuffering)
+		{
+			log_debug("Received %s when transaction is not "
+					  "in buffering mode, ignoring.", line);
+			return true;
+		}
+
+		fformat(replayCtx->txnBuffer, "%s\n", line);
+
+		/* Close replaybuffer to mark it as complete */
+		fclose(replayCtx->txnBuffer);
+
+		/* Stream the transaction buffer into the target database */
+		if (!streamBufferedXID(replayCtx))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		replayCtx->isTranscationBuffering = false;
+
+		/* Remove the symlink to the replay buffer */
+		unlink_file(context->paths.txnlatestfile);
+
+		/* Early exit if we reached the end position */
+		*stop = context->reachedEndPos;
+
+		replayCtx->txnBuffer = NULL;
+	}
+	else
+	{
+		if (replayCtx->txnBuffer != NULL)
```

Comment:
I believe we should use `isTransactionBuffering` here (and fix the typo in the variable name).

Reply:
Ahh, I missed to stage this change! Will do it.